### PR TITLE
8268192: LambdaMetafactory with invokespecial causes VerificationError

### PR DESCRIPTION
**Pull request by dansmithcode**
Small bug fix.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268192](https://bugs.openjdk.java.net/browse/JDK-8268192): LambdaMetafactory with invokespecial causes VerificationError

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4403/head:pull/4403` \
`$ git checkout pull/4403`

Update a local copy of the PR: \
`$ git checkout pull/4403` \
`$ git pull https://git.openjdk.java.net/jdk pull/4403/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4403`

View PR using the GUI difftool: \
`$ git pr show -t 4403`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4403.diff">https://git.openjdk.java.net/jdk/pull/4403.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java [hunk 1]
```diff
@@ -180,8 +180,15 @@ public InnerClassLambdaMetafactory(MethodHandles.Lookup caller,
...
-                              !VerifyAccess.isSamePackage(implClass, implInfo.getDeclaringClass());
+        // If the target class invokes a protected method inherited from a
+        // superclass in a different package, or does 'invokespecial', the
+        // lambda class has no access to the resolved method. Instead, we need
+        // to pass the live implementation method handle to the proxy class
+        // to invoke directly. (javac prefers to avoid this situation by
+        // generating bridges in the target class)
+        useImplMethodHandle = (Modifier.isProtected(implInfo.getModifiers()) &&
+                               !VerifyAccess.isSamePackage(implClass, implInfo.getDeclaringClass())) ||
+                               implKind == H_INVOKESPECIAL;
```
Comment (liach) [inline src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:191]: Won't this make regular private instance method calls use condy as well, as they are invokespecial?

Comment (dansmithcode) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:191]: See this code from the `AbstractValidatingLambdaMetafactory` constructor:

```
            case REF_invokeSpecial:
                // JDK-8172817: should use referenced class here, but we don't know what it was
                this.implClass = implInfo.getDeclaringClass();
                this.implIsInstanceMethod = true;

                // Classes compiled prior to dynamic nestmate support invokes a private instance
                // method with REF_invokeSpecial.
                //
                // invokespecial should only be used to invoke private nestmate constructors.
                // The lambda proxy class will be defined as a nestmate of targetClass.
                // If the method to be invoked is an instance method of targetClass, then
                // convert to use invokevirtual or invokeinterface.
                if (targetClass == implClass && !implInfo.getName().equals("<init>")) {
                    this.implKind = implClass.isInterface() ? REF_invokeInterface : REF_invokeVirtual;
                } else {
                    this.implKind = REF_invokeSpecial;
                }
                break;
```

We turn all same-class invocations into `invokevirtual`. (And all `<init>` invocations have kind `newInvokeSpecial`, mentioning them here is actually useless.)

Comment (dansmithcode) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:191]: Actually, I think I'll fix up this code and the comment...

File: src/java.base/share/classes/java/lang/invoke/AbstractValidatingLambdaMetafactory.java [hunk 2]
```diff
@@ -154,14 +154,10 @@
...
-                //
-                // invokespecial should only be used to invoke private nestmate constructors.
-                // The lambda proxy class will be defined as a nestmate of targetClass.
-                // If the method to be invoked is an instance method of targetClass, then
-                // convert to use invokevirtual or invokeinterface.
-                if (targetClass == implClass && !implInfo.getName().equals("<init>")) {
+                // Classes compiled prior to dynamic nestmate support invoke a private instance
+                // method with REF_invokeSpecial. Newer classes use REF_invokeVirtual or
+                // REF_invokeInterface, and we can use that instruction in the lambda class.
+                if (targetClass == implClass) {
```
Comment (dansmithcode) [inline src/java.base/share/classes/java/lang/invoke/AbstractValidatingLambdaMetafactory.java:160]: The name `<init>` can't appear here. It's only used by `newInvokeSpecial`.

**Review by PaulSandoz — APPROVED** [review]

**Review by mlchung — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**